archiver_fastpath.go

```diff
@@ -0,0 +1,277 @@
+	t, ok := tr.(*temporaryRepository)
+	if !ok {
+		return fmt.Errorf("internal error, not a temporaryRepository")
+	}
+
+	repo := t.Repository
+
+	err := renameReferences(repo, r.ID)
+	if err != nil {
+		return err
+	}
+
+	err = StoreConfig(repo, r)
+	if err != nil {
+		return err
+	}
+
+	err = repo.DeleteRemote("origin")
+	if err != nil {
+		return err
+	}
+
+	rootedRepoCpStart := time.Now()
+	err = copySivaToRemote(ctx, a, ic, t)
+	if err != nil {
+		logger.With(log.Fields{
+			"duration": time.Since(rootedRepoCpStart),
+		}).Errorf(err, "could not copy siva file to FS")
+		return err
+	}
+
+	logger.With(log.Fields{
+		"duration": time.Since(rootedRepoCpStart),
+	}).Debugf("copy siva file to FS")
+
+	return nil
+}
+
+func rootedRefName(
+	name plumbing.ReferenceName,
+	id kallax.ULID,
+) plumbing.ReferenceName {
+	n := fmt.Sprintf("%s/%s", name, id.String())
+	return plumbing.ReferenceName(n)
+}
+
+func renameReferences(repo *git.Repository, id kallax.ULID) error {
+	it, err := repo.References()
+	if err != nil {
+		return err
+	}
+
+	var add []*plumbing.Reference
+	var del []plumbing.ReferenceName
+	err = it.ForEach(func(ref *plumbing.Reference) error {
+		if !strings.HasPrefix(string(ref.Name()), "refs/") {
+			return nil
+		}
+
+		name := rootedRefName(ref.Name(), id)
+
+		var newRef *plumbing.Reference
+		switch ref.Type() {
+		case plumbing.HashReference:
+			newRef = plumbing.NewHashReference(name, ref.Hash())
+		case plumbing.SymbolicReference:
+			target := rootedRefName(ref.Target(), id)
+			newRef = plumbing.NewSymbolicReference(name, target)
+		default:
+			return nil
+		}
+
+		del = append(del, ref.Name())
+		add = append(add, newRef)
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, name := range del {
+		err = repo.Storer.RemoveReference(name)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, ref := range add {
+		err = repo.Storer.SetReference(ref)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func copySivaToRemote(
+	ctx context.Context,
+	a *Archiver,
+	ic model.SHA1,
+	t *temporaryRepository,
+) error {
+	local := a.Copier.Local()
+	origPath := fmt.Sprintf("%s.siva", ic.String())
+	localPath := local.Join(fmt.Sprintf(
+		"%s_%s", ic.String(),
+		strconv.FormatInt(time.Now().UnixNano(), 10),
+	))
+	localSivaPath := filepath.Join(localPath, "siva")
+	localTmpPath := filepath.Join(localPath, "tmp")
+
+	defer util.RemoveAll(local, localPath)
+
+	tmpFs, err := local.Chroot(localTmpPath)
+	if err != nil {
+		return err
+	}
+
+	fs, err := sivafs.NewFilesystem(local, localSivaPath, tmpFs)
+	if err != nil {
+		return err
+	}
+
+	err = RecursiveCopy(t.TempPath, t.TempFilesystem, "/", fs)
+	if err != nil {
+		return err
+	}
+
+	err = fs.Sync()
+	if err != nil {
+		return err
+	}
+
+	return a.Copier.CopyToRemote(ctx, localSivaPath, origPath)
+}
+
+// RecursiveCopy copies a directory to a destination path. It creates all
+// needed directories if destination path does not exist.
+func RecursiveCopy(
```

Comment:
maybe all the `...Copy` functions should at least follow params ordering convention (first is dest, second src):
`Copy(dst Writer, src Reader)` - wdyt?